### PR TITLE
[Merged by Bors] - feature(Analysis/Convex/Topology): convex closed hull

### PR DESCRIPTION
**Pull request by mans0954**
Define the convex closed hull and the absolutely convex closed hull and prove basic properties.

See e.g. Bourbaki, TVS II.13

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mans0954** [comment]
@ADedecker would you like to review this one as well please?

File: Mathlib/Analysis/Convex/Topology.lean [hunk 1]
```diff
@@ -91,6 +91,49 @@ end stdSimplex
...
+
+theorem convex_convexClosedHull {s : Set E} :
+    Convex 𝕜 (convexClosedHull 𝕜 s) := ((convexClosedHull 𝕜).isClosed_closure s).1
+
+theorem isClosed_convexClosedHull {s : Set E} :
+    IsClosed (convexClosedHull 𝕜 s) := ((convexClosedHull 𝕜).isClosed_closure s).2
+
+theorem subset_convexClosedHull {s : Set E} : s ⊆ convexClosedHull 𝕜 s :=
+  (convexClosedHull 𝕜).le_closure s
+
```
Comment (j-loreaux) [inline Mathlib/Analysis/Convex/Topology.lean:118]: I think you also want:
```lean
theorem closure_subset_convexClosedHull {s : Set E} : closure s ⊆ convexClosedHull 𝕜 s :=
  sorry
```

File: Mathlib/Analysis/Convex/Topology.lean [hunk 2]
```diff
@@ -91,6 +91,49 @@ end stdSimplex
...
+theorem convexHull_subseteq_convexClosedHull {s : Set E} :
+    (convexHull 𝕜) s ⊆ (convexClosedHull 𝕜) s :=
+  convexHull_min subset_convexClosedHull convex_convexClosedHull
+
+theorem convexClosedHull_eq_convexClosedHull_closure {s : Set E} :
+    convexClosedHull 𝕜 s = convexClosedHull 𝕜 (closure s) := subset_antisymm
+  (convexClosedHull_min (subset_trans subset_closure subset_convexClosedHull)
+    ⟨convex_convexClosedHull, isClosed_convexClosedHull⟩)
+  (convexClosedHull_min (closure_minimal subset_convexClosedHull isClosed_convexClosedHull)
+    ⟨convex_convexClosedHull, isClosed_convexClosedHull⟩)
```
Comment (j-loreaux) [inline Mathlib/Analysis/Convex/Topology.lean:131]: There should be a much simpler proof using the lemma I just mentioned along the lines of:
```
convexClosedHull 𝕜 s ⊆ convexClosedHull 𝕜 (closure s) ⊆ convexClosedHull 𝕜 (convexClosedHull 𝕜 s) = convexClosedHull 𝕜 s
```
The first is `ClosureOperator.monotone`, the second is the lemma I just mentioned, and the third is `ClosureOperator.idempotent`, which is a `simp` lemma. So, something like (untested) should work:
```suggestion
theorem convexClosedHull_eq_convexClosedHull_closure {s : Set E} :
    convexClosedHull 𝕜 s = convexClosedHull 𝕜 (closure s) :=
  subset_antisymm ((convexClosedHull 𝕜).monotone subset_closure) <| by
    simpa using ((convexClosedHull 𝕜).monotone closure_subset_convexClosedHull)
```

File: Mathlib/Analysis/LocallyConvex/AbsConvex.lean [hunk 3]
```diff
@@ -147,8 +147,65 @@ theorem absConvexHull_nonempty : (absConvexHull 𝕜 s).Nonempty ↔ s.Nonempty
...
+variable {𝕜}
+
+theorem absConvex_convexClosedHull {s : Set E} :
+    AbsConvex 𝕜 (absConvexClosedHull 𝕜 s) := ((absConvexClosedHull 𝕜).isClosed_closure s).1
+
+theorem isClosed_absConvexClosedHull {s : Set E} :
+    IsClosed (absConvexClosedHull 𝕜 s) := ((absConvexClosedHull 𝕜).isClosed_closure s).2
+
+theorem subset_absConvexClosedHull {s : Set E} : s ⊆ absConvexClosedHull 𝕜 s :=
+  (absConvexClosedHull 𝕜).le_closure s
```
Comment (j-loreaux) [inline Mathlib/Analysis/LocallyConvex/AbsConvex.lean:173]: As above, you want the `closure s ⊆ _` lemma.

File: Mathlib/Analysis/LocallyConvex/AbsConvex.lean [hunk 4]
```diff
@@ -147,8 +147,65 @@ theorem absConvexHull_nonempty : (absConvexHull 𝕜 s).Nonempty ↔ s.Nonempty
...
+    (absConvexHull 𝕜) s ⊆ (absConvexClosedHull 𝕜) s :=
+  absConvexHull_min subset_absConvexClosedHull absConvex_convexClosedHull
+
+theorem absConvexClosedHull_eq_absConvexClosedHull_closure {s : Set E} :
+    absConvexClosedHull 𝕜 s = absConvexClosedHull 𝕜 (closure s) := subset_antisymm
+  (absConvexClosedHull_min (subset_trans subset_closure subset_absConvexClosedHull)
+    ⟨absConvex_convexClosedHull, isClosed_absConvexClosedHull⟩)
+  (absConvexClosedHull_min
+    (closure_minimal subset_absConvexClosedHull isClosed_absConvexClosedHull)
+    ⟨absConvex_convexClosedHull, isClosed_absConvexClosedHull⟩)
```
Comment (j-loreaux) [inline Mathlib/Analysis/LocallyConvex/AbsConvex.lean:188]: Here the proof can be simplified the same way.

File: Mathlib/Analysis/Convex/Topology.lean [hunk 5]
```diff
@@ -91,6 +91,49 @@ end stdSimplex
...
+theorem convex_convexClosedHull {s : Set E} :
+    Convex 𝕜 (convexClosedHull 𝕜 s) := ((convexClosedHull 𝕜).isClosed_closure s).1
+
+theorem isClosed_convexClosedHull {s : Set E} :
+    IsClosed (convexClosedHull 𝕜 s) := ((convexClosedHull 𝕜).isClosed_closure s).2
+
+theorem subset_convexClosedHull {s : Set E} : s ⊆ convexClosedHull 𝕜 s :=
+  (convexClosedHull 𝕜).le_closure s
+
+theorem convexClosedHull_min {s t : Set E} : s ⊆ t → Convex 𝕜 t ∧ IsClosed t →
```
[the remark this answers is not in the document]
Comment (mans0954) [reply Mathlib/Analysis/Convex/Topology.lean]: Done, thanks, and similarly for `closedAbsConvexHull_min`.

File: Mathlib/Analysis/Convex/Topology.lean (hunk 1 above)
Comment (mans0954) [reply Mathlib/Analysis/Convex/Topology.lean]: Done, thanks.

File: Mathlib/Analysis/Convex/Topology.lean (hunk 2 above)
Comment (mans0954) [reply Mathlib/Analysis/Convex/Topology.lean]: Done, thanks.

File: Mathlib/Analysis/LocallyConvex/AbsConvex.lean (hunk 3 above)
Comment (mans0954) [reply Mathlib/Analysis/LocallyConvex/AbsConvex.lean]: Done, thanks.

File: Mathlib/Analysis/LocallyConvex/AbsConvex.lean (hunk 4 above)
Comment (mans0954) [reply Mathlib/Analysis/LocallyConvex/AbsConvex.lean]: Done, thanks.

File: Mathlib/Analysis/Convex/Topology.lean [hunk 6]
```diff
@@ -232,6 +275,12 @@ protected theorem Convex.closure {s : Set E} (hs : Convex 𝕜 s) : Convex 𝕜
     (continuous_fst.const_smul _).add (continuous_snd.const_smul _)
   show f x y ∈ closure s from map_mem_closure₂ hf hx hy fun _ hx' _ hy' => hs hx' hy' ha hb hab
 
+theorem convexClosedHull_eq_closure_convexHull {s : Set E} :
+    convexClosedHull 𝕜 s = closure (convexHull 𝕜 s) := subset_antisymm
+  (convexClosedHull_min (subset_trans (subset_convexHull 𝕜 s) subset_closure)
+    ⟨Convex.closure (convex_convexHull 𝕜 s), isClosed_closure⟩)
+  (closure_minimal convexHull_subseteq_convexClosedHull isClosed_convexClosedHull)
```
[the remark this answers is not in the document]
Comment (mans0954) [reply Mathlib/Analysis/Convex/Topology.lean]: Done.

File: Mathlib/Analysis/Convex/Topology.lean [hunk 7]
```diff
@@ -315,6 +268,64 @@ protected theorem Convex.strictConvex {s : Set E} (hs : Convex 𝕜 s)
...
+  (closedConvexHull 𝕜).le_closure s
+
+theorem closure_subset_closedConvexHull {s : Set E} : closure s ⊆ closedConvexHull 𝕜 s :=
+  closure_minimal subset_closedConvexHull isClosed_closedConvexHull
+
+theorem closedConvexHull_min {s t : Set E} (hst : s ⊆ t) (h_conv : Convex 𝕜 t)
+    (h_closed : IsClosed t) : closedConvexHull 𝕜 s ⊆ t :=
+  (closedConvexHull 𝕜).closure_min hst ⟨h_conv, h_closed⟩
+
+theorem convexHull_subseteq_closedConvexHull {s : Set E} :
```
Comment (j-loreaux) [inline Mathlib/Analysis/Convex/Topology.lean:305]: We just write `subset` in the name.
```suggestion
theorem convexHull_subset_closedConvexHull {s : Set E} :
```

File: Mathlib/Analysis/Convex/Topology.lean [hunk 8]
```diff
@@ -315,6 +268,64 @@ protected theorem Convex.strictConvex {s : Set E} (hs : Convex 𝕜 s)
...
+    (convexHull 𝕜) s ⊆ (closedConvexHull 𝕜) s :=
+  convexHull_min subset_closedConvexHull convex_closedConvexHull
+
+theorem closedConvexHull_eq_closedConvexHull_closure {s : Set E} :
+    closedConvexHull 𝕜 s = closedConvexHull 𝕜 (closure s) :=
```
Comment (j-loreaux) [inline Mathlib/Analysis/Convex/Topology.lean:310]: Now that I look at this again, I wonder if it makes more sense turned around as a `simp` lemma?

File: Mathlib/Analysis/LocallyConvex/AbsConvex.lean [hunk 9]
```diff
@@ -158,34 +162,36 @@ variable (𝕜)
...
 
-theorem absConvexClosedHull_eq_absConvexClosedHull_closure {s : Set E} :
-    absConvexClosedHull 𝕜 s = absConvexClosedHull 𝕜 (closure s) := subset_antisymm
-  (absConvexClosedHull_min (subset_trans subset_closure subset_absConvexClosedHull)
-    ⟨absConvex_convexClosedHull, isClosed_absConvexClosedHull⟩)
-  (absConvexClosedHull_min
-    (closure_minimal subset_absConvexClosedHull isClosed_absConvexClosedHull)
-    ⟨absConvex_convexClosedHull, isClosed_absConvexClosedHull⟩)
+theorem closedAbsConvexHull_eq_closedAbsConvexHull_closure {s : Set E} :
+    closedAbsConvexHull 𝕜 s = closedAbsConvexHull 𝕜 (closure s) :=
```
Comment (j-loreaux) [inline Mathlib/Analysis/LocallyConvex/AbsConvex.lean:191]: same here

**Review by j-loreaux — COMMENTED** [review]
Thanks! Please address these comments and wait for CI to finish before merging

bors d+

File: Mathlib/Analysis/Convex/Topology.lean (hunk 8 above)
Comment (mans0954) [reply Mathlib/Analysis/Convex/Topology.lean]: Done, thanks.

File: Mathlib/Analysis/LocallyConvex/AbsConvex.lean (hunk 9 above)
Comment (mans0954) [reply Mathlib/Analysis/LocallyConvex/AbsConvex.lean]: Done, thanks.

**Pull request closed without merging**